cardano-mempool/tests/Test/Cardano/Memory/PoolTests.hs

```diff
@@ -106,85 +108,107 @@ mallocPreFilled preFillByte bc = do
+        touch fmps
+        -- Here we return just the pointers and let the GC collect the ForeignPtrs
+        ptrs <-
+          forM fmps $ \fma ->
+            withForeignPtr fma $ \ptr -> do
+              let bytePtr = castPtr ptr
+              checkBlockBytes block preFillByte bytePtr
+              setPtr bytePtr (blockByteCount block) fillByte
+              pure bytePtr
+        pure (pool, ptrs)
     forM_ ptrs (checkBlockBytes block fillByte)
     checkNumPages pool n numBlocks
     -- Ensure that memory to that the pointers are referencing to is still alive
     touch pool
 
 propPoolAllocateAndFinalize ::
-  forall n.
-  KnownNat n =>
-  Block n ->
-  Positive Int ->
-  Word16 ->
-  Word8 ->
-  Word8 ->
-  Property
+     forall n. KnownNat n
+  => Block n
+  -> Positive Int
+  -> Word16
+  -> Word8
+  -> Word8
+  -> Property
 propPoolAllocateAndFinalize block (Positive n) numBlocks16 emptyByte fullByte =
   monadicIO . run $ do
     let numBlocks = 1 + (fromIntegral numBlocks16 `div` 20)
-    countRef <- newPVar (0 :: Int)
-    chan <- newChan
     pool <-
-      initPool n (mallocPreFilled emptyByte) $ \(ptr :: Ptr (Block n)) -> do
-        setPtr (castPtr ptr) (blockByteCount block) emptyByte
-        void $ atomicAddIntPVar countRef 1
-    -- allocate and finalize blocks concurrently
-    concurrently_
-      (do replicateConcurrently_ numBlocks $ do
-            fp <- grabNextBlock pool
-            withForeignPtr fp (checkBlockBytes block emptyByte)
-            writeChan chan (Just fp)
-          -- place Nothing to indicate we are done allocating blocks
-          writeChan chan Nothing)
-      (fix $ \loop -> do
-         mfp <- readChan chan
-         forM_ mfp $ \fp -> do
-           withForeignPtr fp $ \ptr ->
-             -- fill the newly allocated block
-             setPtr (castPtr ptr) (blockByteCount block) fullByte
-           -- manually finalize every other block and let the GC to pick the rest
-           shouldFinalize <- uniformM globalStdGen
-           when shouldFinalize $ finalizeForeignPtr fp
-           loop)
-    performGC
-    -- allow some time for all blocks to finalize
-    threadDelay 50000
-    -- verify all finalizers have been executed
-    numBlocks' <- atomicReadIntPVar countRef
-    numBlocks' @?= numBlocks
+      ensureAllGCed numBlocks $ \countOneBlockGCed -> do
+        chan <- newChan
+        pool <-
+          initPool n (mallocPreFilled emptyByte) $ \(ptr :: Ptr (Block n)) -> do
+            setPtr (castPtr ptr) (blockByteCount block) emptyByte
+            countOneBlockGCed
+      -- allocate and finalize blocks concurrently
+        pool <$
+          concurrently_
+            (do replicateConcurrently_ numBlocks $ do
+                  fp <- grabNextBlock pool
+                  withForeignPtr fp (checkBlockBytes block emptyByte)
+                  writeChan chan (Just fp)
+              -- place Nothing to indicate we are done allocating blocks
```

Comment:
Indentation looks off

Reply:
Thank you. The hindent automatic formatter I use is pretty buggy and can't handle comments properly.